### PR TITLE
MicStreamSelector full demo and expanded API

### PR DESCRIPTION
**Pull request by andymule**
MicStreamSelector includes XAML demo, and Mic API now includes callback hooks for supporting non-game engine scenarios, like XAML, to let the audio engine drive audio management instead of letting the game engine drive it.

### CODE REVIEW

File: Audio/MicStreamSelector/Projects/MicDemoApp/App.xaml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+﻿<Application
```
Comment (NeerajW) [inline Audio/MicStreamSelector/Projects/MicDemoApp/App.xaml]: Could you please update the README.md with useful information on the Mic API's and the test information?

File: Audio/MicStreamSelector/Projects/MicDemoApp/MicStreamSelector.cs [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+using System.Runtime.InteropServices;
+using System.Diagnostics;
+using Windows.Media.Audio;
+using Windows.Media.Render;
+using Windows.UI.Xaml.Controls;
+using Windows.UI.Xaml;
+
+namespace MicDemoApp
+{
+    public class MicStreamSelector
```
Comment (NeerajW) [inline Audio/MicStreamSelector/Projects/MicDemoApp/MicStreamSelector.cs]: Seems like anyone who wants to use MicStreamSelector will need this wrapper? Should we put this in a better location than here?

File: Audio/MicStreamSelector/Projects/MicDemoApp/MicStreamSelector.cs [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    {
+        // Make sure you enable MusicLibrary and Microphone capabilities in APPX Manifest
```
Comment (NeerajW) [inline Audio/MicStreamSelector/Projects/MicDemoApp/MicStreamSelector.cs]: We need MusicLibrary for saving audio clips?

File: Audio/MicStreamSelector/Projects/MicDemoApp/MicStreamSelector.cs (hunk 2 above)
Comment (davidkline-ms) [inline Audio/MicStreamSelector/Projects/MicDemoApp/MicStreamSelector.cs]: I agree with Neeraj,

Not sure we need to create a new folder structure… yet. What do folks think about putting this in the Utilities\Scripts folder?

**Comment by jwittner** [comment]
The inclusion of the .suo file seems rather odd - any other user specific changes to the solution will result in overwrites of the file and show as changes people probably don't want to commit. Not sure the default startup project nicety is worth that inconvenience in the future.

File: Audio/MicStreamSelector/Projects/MicDemoApp/MicStreamSelector.cs (hunk 3 above)
Comment (andymule) [inline Audio/MicStreamSelector/Projects/MicDemoApp/MicStreamSelector.cs]: I'm open to other places to save audio files. I think it's great to make them easily accessible, but we at least need to make a sub-folder inside Music called "MicStream Recordings" or something similar

**Comment by andymule** [comment]
@jwittner 
it is my understanding that the inclusion of .suo is the only way to ensure that the startup project is the .exe and not one of the two .dll projects, so people can just open the project and hit play without setting the startup project manually first. If there is another way to get the behaviour, we should do it.

File: Audio/MicStreamSelector/Projects/MicDemoApp/App.xaml (hunk 1 above)
Comment (andymule) [inline Audio/MicStreamSelector/Projects/MicDemoApp/App.xaml]: yes, much needed

**Comment by jwittner** [comment]
@andymule I believe that is the only way to set the startup project. The issue I have is that this file is filled with lots of user specific settings and so isn't a good candidate for source control. I don't think the confusion over changes to this file and the potential stomping to other use specific changes that may occur is worth what I see as the rather small gain of having the right project selected on initial opening.

I'd be interested in how the others feel.

File: Audio/MicStreamSelector/Projects/MicDemoApp/MicStreamSelector.cs (hunk 3 above)
Comment (NeerajW) [inline Audio/MicStreamSelector/Projects/MicDemoApp/MicStreamSelector.cs]: Agreed. Please consider this comment resolved.

**Comment by NeerajW** [comment]
@jwittner @andymule as you mentioned, *.suo file is the place for setting startup projects. I believe it's ok to check in this file as part of the project. If someone else changes it, we should be able to see the diff and catch issues if any.

**Comment by jwittner** [comment]
From the documentation on .suo files (emphasis mine): https://msdn.microsoft.com/en-us/library/bb165909.aspx

"The solution user options (.suo) file contains per-user solution options. This file **should not** be checked in to source code control."

**Comment by andymule** [comment]
@jwittner
I have a hunch that making it the first project listed in the Solution will ensure it loads as the default. Will experiment. Will also reach out to VS team, seems like it should be a properly supported solution capability.

**Comment by NeerajW** [comment]
@andymule any updates here?

**Comment by jwittner** [comment]
I lost the .suo debate? =(

**Pull request merged**